### PR TITLE
kv: extract bufferStore from UnionStore, fix #463

### LINKED ISSUE #463: before tidb.runStmt return error, need to rollback the current transaction
In tidb.runStmt, if encount some error in any exec func, you need to rollback the current transaction, otherwise, the txn.UnionStore.Dirty will contain some discard data..

https://github.com/pingcap/tidb/blob/d7c3cc47916db41552b88fb55f2cf0156be4bea1/tidb.go#L171

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

**Comment by disksing** [comment]
PTAL @coocood @ngaut

**Comment by coocood** [comment]
LGTM

File: kv/cache_snapshot.go [hunk 1]
```diff
@@ -98,12 +98,11 @@ func (c *cacheSnapshot) BatchGet(keys []Key) (map[string][]byte, error) {
 	}
 	for _, k := range missedKeys {
 		ks := string(k)
-		if v, ok := values[ks]; ok && len(v) > 0 {
-			err = c.cache.Set(k, v)
+		v, ok := values[ks]
+		if ok && len(v) > 0 {
```
Comment (qiuyesuifeng) [inline kv/cache_snapshot.go]: Seems need continue?

File: kv/kv.go [hunk 2]
```diff
@@ -121,30 +121,50 @@ const (
 
 // Retriever is the interface wraps the basic Get and Seek methods.
 type Retriever interface {
-	// Get gets the value for key k from KV storage.
+	// Get gets the value for key k from kv store.
+	// If corresponding kv pair does not exist, it returns nil and ErrNotExist.
 	Get(k Key) ([]byte, error)
-	// Seek searches for the entry with key k in KV storage.
+	// Seek creates an Iterator positioned on the first entry that k <= entry's key.
+	// If such entry is not found, it returns a invalid Iterator with no error.
```
Comment (qiuyesuifeng) [inline kv/kv.go]: an invalid

Comment (qiuyesuifeng) [inline kv/kv.go]: A little confused why `if such entry is not found, returns no error`?

File: kv/kv.go [hunk 3]
```diff
@@ -121,30 +121,50 @@ const (
...
-// snapshot for read.
-type UnionStore interface {
+// MemBuffer is an in-memory kv collection. It should be released after use.
+type MemBuffer interface {
 	RetrieverMutator
+	// Release releases the buffer.
+	Release()
+}
+
+// BufferStore is the interface that wraps a Retriever for read and a MemBuffer for buffered write.
```
Comment (qiuyesuifeng) [inline kv/kv.go]: Not see `wraps a Retriever` only a MemBuffer wrapped?

File: kv/cache_snapshot.go (hunk 1 above)
Comment (disksing) [inline kv/cache_snapshot.go]: We should cache a nil value if the key does not exist.

File: kv/kv.go (hunk 2 above)
Comment (disksing) [inline kv/kv.go]: It means that the Seek operation is executed with no error, but the returned Iterator is invalid.

File: kv/buffer_store.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/juju/errors"
+)
+
+type bufferStore struct {
+	MemBuffer
+	r Retriever
+}
+
+// NewBufferStore creates a BufferStore using r for read.
+func NewBufferStore(r Retriever) BufferStore {
```
Comment (shenli) [inline kv/buffer_store.go]: Add some comments for bufferStore.

**Comment by disksing** [comment]
PTAL @qiuyesuifeng @shenli

**Pull request merged**